upcloud/resource_upcloud_server.go

```diff
@@ -767,3 +767,111 @@ func serverDefaultTitleFromHostname(hostname string) string {
 	}
 	return fmt.Sprintf("%s%s", hostname, suffix)
 }
+
+func serverValidateHostnameDiagFunc(min, max int) schema.SchemaValidateDiagFunc {
+	return func(v interface{}, path cty.Path) diag.Diagnostics {
+		var diags diag.Diagnostics
+		val, ok := v.(string)
+		if !ok {
+			diags = append(diags, diag.Diagnostic{
+				Severity:      diag.Error,
+				Summary:       "Bad type",
+				Detail:        "expected type to be string",
+				AttributePath: path,
+			})
+			return diags
+		}
+
+		if len(val) < min || len(val) > max {
+			diags = append(diags, diag.Diagnostic{
+				Severity:      diag.Error,
+				Summary:       "Hostname length validation failed",
+				Detail:        fmt.Sprintf("expected hostname length to be in the range (%d - %d), got %d", min, max, len(val)),
+				AttributePath: path,
+			})
+			return diags
+		}
+
+		if err := serverValidateHostname(val); err != nil {
+			diags = append(diags, diag.Diagnostic{
+				Severity:      diag.Error,
+				Summary:       "Hostname validation failed",
+				Detail:        err.Error(),
+				AttributePath: path,
+			})
+		}
+
+		return diags
+	}
+}
+
+// Validate server hostname
+//
+// hostname(7): Each element of the hostname must be from 1 to 63 characters long
+// and the entire hostname, including the dots, can be at most 253 characters long.
+// Valid characters for hostnames are ASCII(7) letters from a to z, the digits from 0 to 9, and the hyphen (-).
+// A hostname may not start with a hyphen.
+//
+// Modified version of isDomainName function from Go's net package (https://pkg.go.dev/net)
+func serverValidateHostname(hostname string) error {
+	const (
+		minLen      int = 1
+		maxLen      int = 253
+		labelMaxLen int = 63
+	)
+	l := len(hostname)
+
+	if l > maxLen || l < minLen {
+		return fmt.Errorf("%s length %d is not in the range %d - %d", hostname, l, minLen, maxLen)
+	}
+
+	if hostname[0] == '.' || hostname[0] == '-' {
+		return fmt.Errorf("%s starts with dot or hyphen", hostname)
+	}
+
+	if hostname[l-1] == '.' || hostname[l-1] == '-' {
+		return fmt.Errorf("%s ends with dot or hyphen", hostname)
+	}
+
+	last := byte('.')
+	nonNumeric := false // true once we've seen a letter or hyphen (either one is required)
+	labelLen := 0
+
+	for i := 0; i < l; i++ {
+		c := hostname[i]
+		switch {
+		case 'a' <= c && c <= 'z' || 'A' <= c && c <= 'Z' || c == '_':
+			nonNumeric = true
+			labelLen++
+		case '0' <= c && c <= '9':
+			labelLen++
+		case c == '-':
+			if last == '.' {
+				return fmt.Errorf("'%s' character before hyphen cannot be dot", hostname[0:i+1])
+			}
+			labelLen++
+			nonNumeric = true
+		case c == '.':
+			if last == '.' || last == '-' {
+				return fmt.Errorf("'%s' character before dot cannot be dot or hyphen", hostname[0:i+1])
+			}
+			if labelLen > labelMaxLen || labelLen == 0 {
+				return fmt.Errorf("'%s' label is not in the range %d - %d", hostname[0:i+1], minLen, labelMaxLen)
```

Comment:
If I'm reading this correctly - this will print the entire hostname, regardless of which label the loop is on. No biggie, but you might want to tweak the wording here a little.

Reply:
Thank you @ka-myl for review!

At this point `hostname[0:i+1]` can also be partial hostname so it might not be label nor hostname :) 

Validation error can something like:
```
Error: Hostname validation failed
 
   with upcloud_server.example,
   on upcloud.tf line XX, in resource "upcloud_server" "example":
   21:   hostname = "example.exampleexampleexampleexampleexampleexampleexampleexampleexampleexample.com"
 
 'example.exampleexampleexampleexampleexampleexampleexampleexampleexampleexample.' label is not in the range 1 - 63
```